### PR TITLE
ENH: Interactive alpha comparison boxplot (D3)

### PR DESCRIPTION
**Pull request by thermokarst**
Redesigned interactive alpha comparison boxplot:
image

Now supports regrouping by metadata category on the fly. Also, visualizer method was renamed to `alpha_group_significance`, as per naming discussion with @gregcaporaso and @jairideout.

### CODE REVIEW

File: q2_diversity/_alpha/_visualizer.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        metadata_category = metadata.get_category(category).to_series()
+        filtered_metadata = metadata_category[alpha_diversity.index]
+        data = pd.concat([alpha_diversity, filtered_metadata], axis=1)
+        names = []
+        groups = []
+        for name, group in data.groupby(filtered_metadata.name):
+            names.append('%s (n=%d)' % (name, len(group)))
+            groups.append(list(group[alpha_diversity.name]))
+
+        if (len(groups) > 1 and len(groups) != len(data)):
```
Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: `len(data.index)`

File: q2_diversity/_alpha/_visualizer.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            filename = 'category-%s.jsonp' % category
```
Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: URLEncode this.

File: q2_diversity/_alpha/_visualizer.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            filenames.append(filename)
+            df = pd.Series(groups, index=names)
```
Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: move into `open` and inline

File: q2_diversity/_alpha/_visualizer.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+            # perform Kruskal-Wallis across all groups
+            kw_H_all, kw_p_all = scipy.stats.mstats.kruskalwallis(*groups)
+            kwAll = json.dumps({'H': kw_H_all, 'p': kw_p_all})
```
Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: move into `open`

File: q2_diversity/_alpha/_visualizer.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                    H, p = scipy.stats.mstats.kruskalwallis(groups[i],
+                                                            groups[j])
+                    kw_H_pairwise.append([names[j], names[i], H, p])
+            kw_H_pairwise = pd.DataFrame(
+                kw_H_pairwise, columns=['Group 1', 'Group 2', 'H', 'p-value'])
+            kw_H_pairwise.set_index(['Group 1', 'Group 2'], inplace=True)
+            kw_H_pairwise['q-value'] = multipletests(
+                kw_H_pairwise['p-value'], method='fdr_bh')[1]
+            kw_H_pairwise.sort_index(inplace=True)
+            outfile = 'kruskal-wallis-pairwise-%s.csv' % category
```
Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: URLEncode

Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: rename this

File: q2_diversity/_alpha/_visualizer.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            outpath = os.path.join(output_dir, outfile)
+            kw_H_pairwise.to_csv(outpath)
+
+            with open(os.path.join(output_dir, filename), 'w') as fh:
+                fh.write('load_data("%s",`' % category)
```
Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: urlencode

File: q2_diversity/_alpha/_visualizer.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                df.to_json(fh, orient='split')
+                fh.write("`,'%s',`" % kwAll)
+                kw_H_pairwise.to_html(fh, classes="table table-striped "
+                                                  "table-hover")
+                fh.write("`,'./%s');" % outfile)
```
Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: Drop relative

File: q2_diversity/_alpha/_visualizer.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+    TEMPLATES = os.path.join(os.path.dirname(os.path.abspath(__file__)),
```
Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: `pkg_resources` check other examples in framework (test plugin base)

File: q2_diversity/_alpha/assets/package.json [hunk 9]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "q2-diversity-alpha-compare",
```
Comment (thermokarst) [inline q2_diversity/_alpha/assets/package.json]: Look into alternatives for name (q2-diversity-alpha) and version (dev?).

File: q2_diversity/_alpha/_visualizer.py (hunk 6 above)
Comment (thermokarst) [inline q2_diversity/_alpha/_visualizer.py]: Look into collapsing multiline string

**Comment by thermokarst** [comment]
Reviewed with @gregcaporaso & @jairideout

File: q2_diversity/_alpha/_visualizer.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+import qiime
+import pandas as pd
+from statsmodels.sandbox.stats.multicomp import multipletests
+from trender import TRender
+
+
+def alpha_group_significance(output_dir: str, alpha_diversity: pd.Series,
+                             metadata: qiime.Metadata) -> None:
+    metadata_df = metadata.to_dataframe()
+    metadata_df = metadata_df.select_dtypes(exclude=[numpy.number])
```
Comment (gregcaporaso) [inline q2_diversity/_alpha/_visualizer.py]: Filter out samples where metadata is empty string in the same way that I'm doing it in `beta_group_significance`:

`metadata = metadata.replace(r'', np.nan).dropna()`

**Pull request merged**